std/net/udp.zig

```diff
@@ -0,0 +1,121 @@
+const std = @import("../index.zig");
+const event = std.event;
+const assert = std.debug.assert;
+const os = std.os;
+const mem = std.mem;
+const net = std.net;
+
+pub fn UDP() type {
+    return struct {
+        channel: *event.Channel(Event),
+        putter: promise,
+        table_lock: event.Lock,
+        fd: i32,
+
+        const Self = this;
+
+        pub const Event = union(enum) {
+            Packet: []u8,
+            Err: Error,
+
+            pub const Error = error{
+                UserResourceLimitReached,
+                SystemResources,
+            };
+        };
+
+        pub fn listen(address: net.Address, loop: *event.Loop, event_buf_count: usize) !*Self {
+            //socket
+            const fd = try os.posixSocket(address.family(), os.posix.SOCK_DGRAM, os.posix.IPPROTO_UDP);
+            errdefer os.close(fd);
+
+            //set non blocking?
+
+            //bind
+            //@breakpoint();
+            try os.posixBind(fd, address);
+
+            const channel = try event.Channel(Self.Event).create(loop, event_buf_count);
+            errdefer channel.destroy();
+
+            var result: *Self = undefined;
+            _ = try async<loop.allocator> eventPutter(fd, channel, &result);
+            return result;
+        }
+
+        pub fn destroy(self: *Self) void {
+            cancel self.putter;
+        }
+
+        pub async fn write(self: *Self, file_path: []const u8, value: V) !?V {
+        }
+
+        async fn eventPutter(fd: i32, channel: *event.Channel(Event), out: **Self) void {
+            // TODO https://github.com/ziglang/zig/issues/1194
+            suspend {
+                resume @handle();
+            }
+
+            const loop = channel.loop;
+
+            var udp = Self{
+                .putter = @handle(),
+                .channel = channel,
+                .table_lock = event.Lock.init(loop),
+                .fd = fd,
+            };
+            out.* = &udp;
+
+            loop.beginOneEvent();
+
+            defer {
+                loop.finishOneEvent();
+                os.close(fd);
+                channel.destroy();
+            }
+
+            var event_buf: [1500]u8 = undefined;
+
+            while (true) {
+                const rc = os.darwin.read(fd, &event_buf, event_buf.len);
+                const errno = os.darwin.getErrno(rc); //maybe this shouldn't be called so much?
```

Comment:
can you explain this comment?

Reply:
good stuff!

Reply:
This is a bit weird. Or maybe the errnoWrap function. Because read returns the number of bytes read iirc or an error code when it's <0. The errnoWrap makes it always return something positive and the only way to figure out if it's an error is by calling getErrno?

Maybe these errors should actually be errors and not some vague return value magic?

(assuming this works the same as Linux, otherwise discard my comment. It's still weird though imo.)